### PR TITLE
SC-206 ~ Add basic LibreOffice Online Client

### PR DESCRIPTION
**Pull request by pinussilvestrus**
**Waiting for our LibreOffice instance to be working....** Checkout https://ticketsystem.schul-cloud.org/browse/SC-206 for more details.

You don't have to checkout the corresponding server-branch (https://github.com/schul-cloud/schulcloud-server/pull/303) since we have deployed it otherwise, but you have to set the following environment variable to display:

`export LIBRE_OFFICE_CLIENT_URL=http://ecs-80-158-4-11.reverse.open-telekom-cloud.com:9980/`

For now I've just implemented a rudimentary way to open office files (Word, PowerPoint, text) in our LibreOffice Online instance (as a frame). This is the basic workflow for integrating wopi-infrastructures (https://wopi.readthedocs.io/en/latest/overview.html#integration-process)

image

image

image

(if no env is set)
image

There is no design involved for now. We can discuss, how the open-link (which is only a pure link below the other stuff right now) can be integrated in the file-card. Maybe @adrianjost @janrenz @Langleu  have ideas for it?

### CODE REVIEW

**Comment by pinussilvestrus** [comment]
image

**Comment by pinussilvestrus** [comment]
Aktuelle Probleme:

- Textdateien öffnen sich zwar, aber sind nicht editierbar (read-only?)
- docx/xlsx/...-Dateien öffnen sich gar nicht (Fehlermeldung):

image

- doc...-Dateien öffnen sich zwar, aber nur kryptisch:

image

**Comment by Langleu** [comment]
progress is mostly updated in https://github.com/schul-cloud/schulcloud-server/pull/303 as main source

**Comment by Langleu** [comment]
to test use the online instance:
`LIBRE_OFFICE_CLIENT_URL=http://ecs-80-158-4-11.reverse.open-telekom-cloud.com:9980/` `BACKEND_URL=http://ecs-80-158-4-11.reverse.open-telekom-cloud.com:9000/`
as env variables when starting the client, server then obv not needed.

New File Creation possible (docx, xlsx, pptx)

image
image
---

File Sharing had to be adjusted a bit for libre office, as we don't want the file directly downloaded
image
Bearbeiten => user ends up in libre office
Runterladen => user downloads file
==> both cases the file is being shared with the user and in both cases the user can afterwards access libre office

---

Course Files:
Any file that can be opened in libre Office is restricted for students and have only view rights, but not write rights.
image
Teachers can toggle write access for students with the new button on the right.

---

If a user opens a file in libre office he will be redirected to a new page with iframe.
Some applications cause errors or weird behaviour in an iframe, under each iframe is some sort of hint through which libre office can be opened in an external tab.

image

**Comment by janrenz** [comment]
Statt Dateiendung wurde ich das Dateityp nennen und als Erstes auswählen lassen ...

**Comment by pinussilvestrus** [comment]
@Langleu I'm kinda proud of what you made my with my basic wopi-structure. Looks very cool ! 👍

**Comment by janrenz** [comment]
@adrianjost Can you provide a screenshot of the themed version?

**Comment by adrianjost** [comment]
habe nun eine theme Datei hinzugefügt. Muss aber noch eingebunden werden. Darum kümmert sich Lars am Donnerstag.

# SC Theme
image

# N21 Theme
image

**Comment by janrenz** [comment]
looks nice.

**Comment by Langleu** [comment]
Habe das Theming wieder entfernt und ausgelagert in den Docker Container.
https://github.com/schul-cloud/libreoffice-online-builder/pull/3

Es ist einfacher, wenn man die benötigten CSS/Images direkt im LibreOffice hat und habe diese entsprechend beim building/bundling hinzugefügt.

Man hätte es natürlich auch so lassen können, aber wäre auf Dauer nicht so persistent wegen Docker.

File: controllers/files.js [hunk 1]
```diff
@@ -223,6 +255,49 @@ const registerSharedPermission = (userId, filePath, shareToken, req) => {
...
+ */
+const getLibreOfficeUrl = (fileId, accessToken) => {
+    if (!process.env.LIBRE_OFFICE_CLIENT_URL) {
+        logger.error('LibreOffice env is currently not defined.');
+        return;
+    }
+
+    // in the form like: http://ecs-80-158-4-11.reverse.open-telekom-cloud.com:9980/
+    const libreOfficeBaseUrl = process.env.LIBRE_OFFICE_CLIENT_URL;
+    const wopiRestUrl = process.env.BACKEND_URL || 'http://localhost:3030/';
```
Comment (kremer-io) [inline controllers/files.js]: We should use `PUBLIC_BACKEND_URL` here, shouldn't we?

Comment (Langleu) [reply controllers/files.js]: you are right :+1:

**Comment by Langleu** [comment]
@kremer-io I don't think so. Docx/pptx/xlsx is also supported by libreoffice (online/offline) and I don't see the advantage of providing another file type if the other ones are equal if not even superior, especially as educational institutions are able to get the microsoft office suite for free or almost free.

**Comment by kremer-io** [comment]
@Langleu Ok.

**Comment by kremer-io** [comment]
big thx at @pinussilvestrus and @Langleu

**Pull request merged**